### ISSUE TITLE
Assertion failed: sps->empty () bpkg-0.17.0-a.0.20240423105141.40bb034f57cf\bpkg\pkg-build.cxx, line 6879

### ISSUE THREAD

**Issue by KarlHedlund** [build2/build2#382]
I'm afraid I can't provide a reproducer but what may be the issue when this assertion fails? It is a project where I'm initializing a new configuration after having already initialized and built another configuration.

```
trace: execute_plan: simulate: yes
trace: pkg_configure: src_root: C:\hello\views\libviews\, out_root: C:\hello-build\x86_64-windows-msvc-debug-external\libviews\
trace: pkg_configure: src_root: C:\hello\views\demo\, out_root: C:\hello-build\x86_64-windows-msvc-debug-external\demo\
Assertion failed: sps->empty (), file C:\Users\user\Downloads\build2-toolchain-0.17-a.0\bpkg-0.17.0-a.0.20240423105141.40bb034f57cf\bpkg\pkg-build.cxx, line 6879
error: process bpkg terminated abnormally: aborted
rmdir -r C:\hello\.bdep\tmp\
```

**Comment by boris-kolpackov** [comment]
Can you provide the complete set of the command lines you used to trigger this? This will at least give us a sense of the usage scenario we are looking at.

**Comment by helmesjo** [comment]
Above was run with our `init` bash script, which does the project/external/host split (eg. so that we don't need to fetch and rebuild all external dependencies each clean run on our CI).

This is the basic setup (skipping checks if a config already exists in `.bdep` and if it should be removed first):
```bash
# BUILD_DIR         (eg. libhello-build/gcc)
# BUILD_DIR_EXT     (eg. libhello-build/gcc-external)
# BUILD_DIR_HOST    (eg. libhello-build/gcc-host)
# CONFIG_NAME       (eg. gcc.build)
# CONFIG_NAME_EXT   (eg. gcc-external.build)
# CONFIG_NAME_HOST  (eg. gcc-host.build)
# PACKAGES          (a list of all our packages in project being initialized)

# We use fixed UUIDs for configurations to avoid "mismatch" errors if old ones needs to be recreated
# The UUIDs are generated from the hashed (md5sum, 128-bit) config names
format_str_to_uuid(){
    hash=$(echo $1 | md5sum)
    uuid_formatted=$(echo "${hash:0:8}-${hash:8:4}-${hash:12:4}-${hash:16:4}-${hash:20:12}")
    echo $uuid_formatted
}

# Empty project
bdep init \
    --empty \
    --directory "$PROJECT_DIR"

# Always create fresh @<config> (to avoid conflicts)
bpkg cfg-create \
    --uuid $(format_str_to_uuid $CONFIG_NAME) \
    --name $CONFIG_NAME \
    --directory "$BUILD_DIR" \
    --wipe \
    cc

# Create @<config>-external
bpkg cfg-create \
    --uuid $(format_str_to_uuid $CONFIG_NAME_EXT) \
    --name $CONFIG_NAME_EXT \
    --directory "$BUILD_DIR_EXT" \
    --wipe \
    cc

# Create @host
bpkg cfg-create \
    --uuid $(format_str_to_uuid $CONFIG_NAME_HOST) \
    --name $CONFIG_NAME_HOST \
    --directory "$BUILD_DIR_HOST" \
    --type host \
    --wipe \
    cc \
    config.config.load=~host

# Link <config> to <config>-external (so dependencies can be resolved from either)
bpkg cfg-link \
    --directory "$BUILD_DIR" "$BUILD_DIR_EXT" \
    --relative

# Add @host to project
bdep config add \
    --directory "$PROJECT_DIR" \
    @$CONFIG_NAME_HOST "$BUILD_DIR_HOST" \
    --no-default \
    --forward

# Add @<config>-external to project
bdep config add \
    --directory "$PROJECT_DIR" \
    @$CONFIG_NAME_EXT "$BUILD_DIR_EXT" \
    --no-default \
    --no-forward

# Add @<config> to project
bdep config add \
    --directory "$PROJECT_DIR" \
    @$CONFIG_NAME "$BUILD_DIR" \
    --no-default \
    --no-forward

# Make config being initialized default & forwarded
bdep config set \
  --directory "$PROJECT_DIR" \
  @$CONFIG_NAME \
  --default \
  --forward

# Initialize in <config>-external
bdep init \
    @${CONFIG_NAME_EXT} \
    --directory "$PROJECT_DIR"

# Drop & deinitialize local packages from <config>-external
bdep deinit \
    @${CONFIG_NAME_EXT} \
    --directory "$PROJECT_DIR" \
    --force
bpkg pkg-drop \
    ${PACKAGES[@]} \
    --directory "$BUILD_DIR_EXT" \
    --keep-unused \
    --yes

# Initialize in <config>, which will now use external packages from <config>-external
bdep init @$CONFIG_NAME \
    --directory "$PROJECT_DIR"

bdep sync --upgrade --yes
```

**Comment by boris-kolpackov** [comment]
Thanks for the details. Is that last `bdep-sync` that triggered the assert? Any changes to the state of things (e.g., new version became available, etc) that might be relevant (presumable this didn't happen before so there must be some trigger)?

**Comment by KarlHedlund** [comment]
No it is the last bdep init, the one under the:
`# Initialize in <config>, which will now use external packages from <config>-external`

It seems to be something related to having multiple configurations. I've cleaned everything and initialized from scratch, and so far it has always been the second configuration being initialized that fails. We first encountered this having built a release configuration for quite some time and then trying building the debug configuration.

However, I cleaned again and started with the debug configuration this time and then it works, but when I try to initialize the release configuration that fails instead this time.

I can't think of any particular changes externally, as I have tested this over and over in various ways in a short period of time and had the same consistent results.
